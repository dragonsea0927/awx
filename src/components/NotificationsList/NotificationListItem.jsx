import React from 'react';
import { shape, number, string, bool, func } from 'prop-types';
import { I18n } from '@lingui/react';
import { t } from '@lingui/macro';
import { Link } from 'react-router-dom';
import {
  Badge,
  Switch,
  DataListItem,
  DataListItemRow,
  DataListItemCells,
  DataListCell,
} from '@patternfly/react-core';

function NotificationListItem (props) {
  const {
    canToggleNotifications,
    notification,
    detailUrl,
    successTurnedOn,
    errorTurnedOn,
    toggleNotification
  } = props;
  const capText = {
    textTransform: 'capitalize'
  };
  const toggleCellStyles = {
    display: 'flex',
    justifyContent: 'flex-end',
  };

  return (
    <I18n>
      {({ i18n }) => (
        <DataListItem
          aria-labelledby={`items-list-item-${notification.id}`}
          key={notification.id}
        >
          <DataListItemRow>
            <DataListItemCells dataListCells={[
              <DataListCell key="name">
                <Link
                  to={{
                    pathname: detailUrl
                  }}
                  style={{ marginRight: '1.5em' }}
                >
                  <b id={`items-list-item-${notification.id}`}>{notification.name}</b>
                </Link>
                <Badge
                  style={capText}
                  isRead
                >
                  {notification.notification_type}
                </Badge>
              </DataListCell>,
              <DataListCell key="toggles" style={toggleCellStyles}>
                <Switch
                  id={`notification-${notification.id}-success-toggle`}
                  label={i18n._(t`Successful`)}
                  isChecked={successTurnedOn}
                  isDisabled={!canToggleNotifications}
                  onChange={() => toggleNotification(
                    notification.id,
                    successTurnedOn,
                    'success'
                  )}
                  aria-label={i18n._(t`Toggle notification success`)}
                />
                <Switch
                  id={`notification-${notification.id}-error-toggle`}
                  label={i18n._(t`Failure`)}
                  isChecked={errorTurnedOn}
                  isDisabled={!canToggleNotifications}
                  onChange={() => toggleNotification(
                    notification.id,
                    errorTurnedOn,
                    'error'
                  )}
                  aria-label={i18n._(t`Toggle notification failure`)}
                />
              </DataListCell>
            ]}
            />
          </DataListItemRow>
        </DataListItem>
      )}
    </I18n>
  );
}

NotificationListItem.propTypes = {
  notification: shape({
    id: number.isRequired,
    name: string.isRequired,
    notification_type: string.isRequired,
  }).isRequired,
  canToggleNotifications: bool.isRequired,
  detailUrl: string.isRequired,
  errorTurnedOn: bool,
  successTurnedOn: bool,
  toggleNotification: func.isRequired,
};

NotificationListItem.defaultProps = {
  errorTurnedOn: false,
  successTurnedOn: false,
};

export default NotificationListItem;
