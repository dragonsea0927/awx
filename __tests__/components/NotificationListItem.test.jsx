import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { I18nProvider } from '@lingui/react';
import NotificationListItem from '../../src/components/NotificationsList/NotificationListItem';

describe('<NotificationListItem />', () => {
  let wrapper;
  const toggleNotification = jest.fn();

  afterEach(() => {
    if (wrapper) {
      wrapper.unmount();
      wrapper = null;
    }
  });

  test('initially renders succesfully', () => {
    wrapper = mount(
      <I18nProvider>
        <MemoryRouter>
          <NotificationListItem
            itemId={9000}
            toggleNotification={toggleNotification}
            detailUrl="/foo"
            notificationType="slack"
          />
        </MemoryRouter>
      </I18nProvider>
    );
    expect(wrapper.length).toBe(1);
  });

  test('handles success click when toggle is on', () => {
    wrapper = mount(
      <I18nProvider>
        <MemoryRouter>
          <NotificationListItem
            itemId={9000}
            successTurnedOn
            toggleNotification={toggleNotification}
            detailUrl="/foo"
            notificationType="slack"
          />
        </MemoryRouter>
      </I18nProvider>
    );
    wrapper.find('Switch').first().find('input').simulate('change');
    expect(toggleNotification).toHaveBeenCalledWith(9000, true, 'success');
  });

  test('handles success click when toggle is off', () => {
    wrapper = mount(
      <I18nProvider>
        <MemoryRouter>
          <NotificationListItem
            itemId={9000}
            successTurnedOn={false}
            toggleNotification={toggleNotification}
            detailUrl="/foo"
            notificationType="slack"
          />
        </MemoryRouter>
      </I18nProvider>
    );
    wrapper.find('Switch').first().find('input').simulate('change');
    expect(toggleNotification).toHaveBeenCalledWith(9000, false, 'success');
  });

  test('handles error click when toggle is on', () => {
    wrapper = mount(
      <I18nProvider>
        <MemoryRouter>
          <NotificationListItem
            itemId={9000}
            errorTurnedOn
            toggleNotification={toggleNotification}
            detailUrl="/foo"
            notificationType="slack"
          />
        </MemoryRouter>
      </I18nProvider>
    );
    wrapper.find('Switch').at(1).find('input').simulate('change');
    expect(toggleNotification).toHaveBeenCalledWith(9000, true, 'error');
  });

  test('handles error click when toggle is off', () => {
    wrapper = mount(
      <I18nProvider>
        <MemoryRouter>
          <NotificationListItem
            itemId={9000}
            errorTurnedOn={false}
            toggleNotification={toggleNotification}
            detailUrl="/foo"
            notificationType="slack"
          />
        </MemoryRouter>
      </I18nProvider>
    );
    wrapper.find('Switch').at(1).find('input').simulate('change');
    expect(toggleNotification).toHaveBeenCalledWith(9000, false, 'error');
  });
});
