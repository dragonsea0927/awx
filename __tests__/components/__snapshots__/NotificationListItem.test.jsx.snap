// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotificationListItem canToggleNotifications /> initially renders succesfully 1`] = `
<NotificationListItem
  canToggleNotifications={true}
  detailUrl="/foo"
  errorTurnedOn={false}
  notification={
    Object {
      "id": 9000,
      "name": "Foo",
      "notification_type": "slack",
    }
  }
  successTurnedOn={false}
  toggleNotification={[MockFunction]}
>
  <I18n
    update={true}
    withHash={true}
  >
    <DataListItem
      aria-labelledby="items-list-item-9000"
      className=""
      isExpanded={false}
      key="9000"
    >
      <li
        aria-labelledby="items-list-item-9000"
        className="pf-c-data-list__item"
      >
        <DataListItemRow
          className=""
          key=".0"
          rowid="items-list-item-9000"
        >
          <div
            className="pf-c-data-list__item-row"
          >
            <DataListItemCells
              className=""
              dataListCells={
                Array [
                  <DataListCell
                    alignRight={false}
                    className=""
                    isFilled={true}
                    isIcon={false}
                    width={1}
                  >
                    <Link
                      replace={false}
                      style={
                        Object {
                          "marginRight": "1.5em",
                        }
                      }
                      to={
                        Object {
                          "pathname": "/foo",
                        }
                      }
                    >
                      <b
                        id="items-list-item-9000"
                      >
                        Foo
                      </b>
                    </Link>
                    <Badge
                      className=""
                      isRead={true}
                      style={
                        Object {
                          "textTransform": "capitalize",
                        }
                      }
                    >
                      slack
                    </Badge>
                  </DataListCell>,
                  <DataListCell
                    alignRight={false}
                    className=""
                    isFilled={true}
                    isIcon={false}
                    style={
                      Object {
                        "display": "flex",
                        "justifyContent": "flex-end",
                      }
                    }
                    width={1}
                  >
                    <Switch
                      aria-label="Toggle notification success"
                      className=""
                      id="notification-9000-success-toggle"
                      isChecked={false}
                      isDisabled={false}
                      label="Successful"
                      onChange={[Function]}
                    />
                    <Switch
                      aria-label="Toggle notification failure"
                      className=""
                      id="notification-9000-error-toggle"
                      isChecked={false}
                      isDisabled={false}
                      label="Failure"
                      onChange={[Function]}
                    />
                  </DataListCell>,
                ]
              }
              key=".0"
              rowid="items-list-item-9000"
            >
              <div
                className="pf-c-data-list__item-content"
              >
                <DataListCell
                  alignRight={false}
                  className=""
                  isFilled={true}
                  isIcon={false}
                  key="name"
                  width={1}
                >
                  <div
                    className="pf-c-data-list__cell"
                  >
                    <Link
                      replace={false}
                      style={
                        Object {
                          "marginRight": "1.5em",
                        }
                      }
                      to={
                        Object {
                          "pathname": "/foo",
                        }
                      }
                    >
                      <a
                        onClick={[Function]}
                        style={
                          Object {
                            "marginRight": "1.5em",
                          }
                        }
                      >
                        <b
                          id="items-list-item-9000"
                        >
                          Foo
                        </b>
                      </a>
                    </Link>
                    <Badge
                      className=""
                      isRead={true}
                      style={
                        Object {
                          "textTransform": "capitalize",
                        }
                      }
                    >
                      <span
                        className="pf-c-badge pf-m-read"
                        style={
                          Object {
                            "textTransform": "capitalize",
                          }
                        }
                      >
                        slack
                      </span>
                    </Badge>
                  </div>
                </DataListCell>
                <DataListCell
                  alignRight={false}
                  className=""
                  isFilled={true}
                  isIcon={false}
                  key="toggles"
                  style={
                    Object {
                      "display": "flex",
                      "justifyContent": "flex-end",
                    }
                  }
                  width={1}
                >
                  <div
                    className="pf-c-data-list__cell"
                    style={
                      Object {
                        "display": "flex",
                        "justifyContent": "flex-end",
                      }
                    }
                  >
                    <Switch
                      aria-label="Toggle notification success"
                      className=""
                      id="notification-9000-success-toggle"
                      isChecked={false}
                      isDisabled={false}
                      label="Successful"
                      onChange={[Function]}
                    >
                      <label
                        className="pf-c-switch"
                        htmlFor="notification-9000-success-toggle"
                      >
                        <input
                          aria-label="Toggle notification success"
                          checked={false}
                          className="pf-c-switch__input"
                          disabled={false}
                          id="notification-9000-success-toggle"
                          onChange={[Function]}
                          type="checkbox"
                        />
                        <span
                          className="pf-c-switch__toggle"
                        />
                        <span
                          aria-hidden="true"
                          className="pf-c-switch__label pf-m-on"
                        >
                          Successful
                        </span>
                        <span
                          aria-hidden="true"
                          className="pf-c-switch__label pf-m-off"
                        >
                          Successful
                        </span>
                      </label>
                    </Switch>
                    <Switch
                      aria-label="Toggle notification failure"
                      className=""
                      id="notification-9000-error-toggle"
                      isChecked={false}
                      isDisabled={false}
                      label="Failure"
                      onChange={[Function]}
                    >
                      <label
                        className="pf-c-switch"
                        htmlFor="notification-9000-error-toggle"
                      >
                        <input
                          aria-label="Toggle notification failure"
                          checked={false}
                          className="pf-c-switch__input"
                          disabled={false}
                          id="notification-9000-error-toggle"
                          onChange={[Function]}
                          type="checkbox"
                        />
                        <span
                          className="pf-c-switch__toggle"
                        />
                        <span
                          aria-hidden="true"
                          className="pf-c-switch__label pf-m-on"
                        >
                          Failure
                        </span>
                        <span
                          aria-hidden="true"
                          className="pf-c-switch__label pf-m-off"
                        >
                          Failure
                        </span>
                      </label>
                    </Switch>
                  </div>
                </DataListCell>
              </div>
            </DataListItemCells>
          </div>
        </DataListItemRow>
      </li>
    </DataListItem>
  </I18n>
</NotificationListItem>
`;
